src/java.net.http/share/classes/java/net/http/HttpResponse.java

```diff
@@ -143,6 +144,14 @@ public interface HttpResponse<T> {
      */
     public T body();
 
+    /**
+     * Returns the body if the given predicate is satisfied. 
+     * 
+     * @param predicate the Predicate to test
+     * @return an Optional containing the response body if the predicate returns true
+     */
+    public Optional<T> bodyWhen(Predicate<ResponseInfo> predicate) throws IOException;
```

Comment:
This is a neat idea which will allow to append additional steps like `.orElseThrows(...)` etc...

Reply:
I wonder if it should be a `Predicate<HttpResponse>` instead?

Reply:
No, we added `HttpResponse.ResponseInfo` for this kind of purpose. It's an immutable subset of HttpResponse state (notably excluding the body itself which is potentially mutable). Also it would be strange to make the body available to a predicate that is deciding whether to provide access to the body.

Reply:
OK - I see the point about the body. I was asking because `HttpResponse` has some additional info like request, previous request, and SSL session. On the other hand all what you need to make a decision here is probably status code and response headers which `ResponseInfo` provides.